lib/splitclient-rb/cache/senders/impressions_sender.rb

```diff
@@ -31,13 +31,19 @@ def call
         private
 
         def impressions_thread
-          Thread.new do
-            @config.logger.info('Starting impressions service')
+          @config.threads[:impressions_sender] = Thread.new do
+            begin
+              @config.logger.info('Starting impressions service')
 
-            loop do
+              loop do
+                post_impressions
+
+                sleep(SplitIoClient::Utilities.randomize_interval(@config.impressions_refresh_rate))
+              end
+            rescue SplitIoClient::ImpressionShutdownException
               post_impressions
 
-              sleep(SplitIoClient::Utilities.randomize_interval(@config.impressions_refresh_rate))
+              @impressions_repository.clear
```

Comment:
You are accounting for the new impressions listener shutdown as well?

Reply:
@maksimf I'm merging this. Mind addressing the impression listener shutdown right now? I'm merging this now.